### DISCUSSION TITLE
Welcome to CG-SQL-author Discussions!

### DISCUSSION THREAD

**Discussion by ricomariani** [ricomariani/CG-SQL-author#7] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ricomariani** [comment]
Enabled discussions by request.  This is probably better than having a blog for announcements also.

**Comment by grifx** [comment]
Hi Rico, Hi everyone

Thanks for enabling Discussions.

I'm Joris from Sydney. I'm a software engineer currently working for a local consumer opinion website in Australia. Over the years, I developed interest in Domain-Driven Design (DDD), Command Query Responsibility Segregation (CQRS), and Event Sourcing (ES). Recently I discovered a pattern used in Industry 4.0 called "Unified Namespace". It made me rethink about the way I was persisting/fetching/sharding data and I developed this idea of "Unified Tenant Namespace". Imagine a fuse FS that would provide views (file) for each tenant (folder). Long story short, it made me consider using a SQLite (for each tenant). My research for tools to manipulate SQLite dbs lead me to CG-SQL. I did some rough benchmarks, rebuilding a read db with SQLite and CG-SQL from domain events is extremely performant. Using a single sqlite file per tenant offers many other benefits like canary releases.

I've always tried to avoid ORMs. I find them too heavy, too slow, too annoying to optimize. They don't provide mechanism to compose different kinds of data sources to serve a unified model (like GraphQL) which translates to adhoc glue code.
Many years ago, after building simple tools to compose data sources, data enrichers / transformers. I realised ORMs were not that useful so I completely stopped using them. Since I discovered CG-SQL I kept trying to think about ways to leverage it to improve both data fetching and persistence of my READ model.

What I like about it:
- It allows us to build typescript types from Json outputs
- Type system
- Using control flow to filter / fetch similar tables
- Allows composition with economical joins

What I would like CG-SQL to do:
- Return multi dimensional structures to avoid hydration from flat structures and useless iteration/copy of data due to one-to-many relationships
- it looks like we can yield object set
  - firstname, lastname, job
  - "John",  "Doe",  {"title": "driver", "company": "uber"}
- It would be nice to yield object list and nested objects
  - firstname, lastname, jobs
  - "John",  "Doe",  [{"title": "driver", "company": "uber"}, {"title": "developer", "company": "google"}]

Perhaps it's not a good idea but the lack of example makes it hard for me to build a strong coherent mental model.

Feedback on the project:
- It's not clear that the project is not dead and that someone cares about making it work for future versions of SQLite.
- Lack of a comprehensive example. Maybe a tiny LUA app
- In general I find examples a bit too abstract `t(a, b, c, d)` vs `post(id, title, body, createdAt)`

Joris

**Comment by ricomariani** [comment]
One of the last things I did for this project while I was at meta was to add parent/child result set combos so that you could make more flexible result sets.  I added some examples for that and I could be convinced to add more examples of other things pretty easily.  Of course wiki and document contributions are always welcome.  

I have not added support for arbitrary objects in the results sets because it would make the ABI a bit crazy.  However, I gave a lot of thought to adding native support for dictionaries as an object type which would give you a lot of flexibility.  Even without that though, nested result sets are very powerful.  

It turns out dictionaries add a lot of language complexity -- not because of the runtime support for them, that's trivial, but because the data type of the dictionary would have to be some flexible thing.  If you make them have fixed schema you get something like a cursor and an interface that is astonishingly like a result set with one row (you can already generate those with the OUT statement).  If you give them flexible schema (which we did at Meta) then you can accomplish it by writing helper functions to create a dictionary object and return it and add `integer`, `text`, `bool`, and `real` to it.  But the object is opaque to the CQL language.

Anyway, I've given a lot of thought to dictionaries... it was the next thing I was going to tackle before I got laid off.

But there's nothing stopping you from returning say object<dict> from a result set, you can't get one from SQL but you can create out and then OUT or OUT UNION a loaded cursor with one in it.  If the ABI issues aren't a problem for your system you can just do it.  If I did it  it would end up looking something like `object<your_shape dict>` with helpers to create one probably from a cursor and update one.

There are already dictionary helpers in the runtime such as `cql_string_dictionary_create()`.  I could write a sample illustrating how to use one.  As it happens there is already code like that in the tests.  You can search for `let dict := cql_string_dictionary_create();` in run_test.sql.  Something a bit more flexible than this is required but it's very doable.

**Comment by ricomariani** [comment]
It looks like all I have to do is make a version of `cql_hashtab` that allows heterogeneous value types.  The code would be mostly the same.  That could be a good starting place for object support.  It just needs a type code for each entry.

**Comment by ricomariani** [comment]
I added another example that shows how to use nested result sets to effectively create records.

https://github.com/ricomariani/CG-SQL-author/wiki/Rowset-with-embedded-objects-for-flexible-columns

Note that most of the code for managing such result sets is shared and the types are just constant metadata so these are not that expensive.

**Reply by grifx** [reply]
Thank you! I need to digest a lot of concepts now :D. I'll try to build a complex json from a procedure while having economical joins.

**Reply by ricomariani** [reply]
The rowsets themselves nest so in principle there is nothing you can't express this way.  This might be the cleanest way to get what you want really.  Adding dictionaries also has value but it just adds yet another way to do almost the same thing.

Note that if you are using a build of SQLite with the JSON support internally you can access those functions by telling SQLite they exist using `declare select function` just as though they were a UDF.  You may need to use `declare select function foo no check` if you need varargs.
